### PR TITLE
feat(util-dynamodb): enable undefined values removal in marshall

### PR DESCRIPTION
**Pull request by trivikr**
*Issue #, if available:*
Fixes: https://github.com/aws/aws-sdk-js-v3/issues/1816

*Description of changes:*
Enables removal of undefined values in marshall by setting `removeUndefinedValues=true`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/util-dynamodb/src/convertToAttr.ts [hunk 1]
```diff
@@ -22,44 +22,52 @@ export const convertToAttr = (data: NativeAttributeValue, options?: marshallOpti
 };
 
 const convertToListAttr = (data: NativeAttributeValue[], options?: marshallOptions): { L: AttributeValue[] } => ({
-  L: data.map((item) => convertToAttr(item, options)),
+  L: data
+    .filter((item) => !options?.removeUndefinedValues || (options?.removeUndefinedValues && item !== undefined))
```
Comment (AllanZhengYP) [inline packages/util-dynamodb/src/convertToAttr.ts:26]: Can the undefined check be moved to `convertToAttr`? Instead of under each convert functions.

Comment (trivikr) [reply packages/util-dynamodb/src/convertToAttr.ts:26]: I tried to attempt it in first commit [`3578da6` (#1840)](https://github.com/aws/aws-sdk-js-v3/pull/1840/commits/3578da69049dd5cf8357ca9197a59ff394f58a4d)

The issue is that `convertToAttr` is exported and should throw if `undefined` is passed to it.
Let me re-attempt it to reduce duplication, and post-process undefined value returned by convertToAttr as PR is now in a good shape.

Comment (trivikr) [reply packages/util-dynamodb/src/convertToAttr.ts:26]: The issue is `convertToAttr` is expected to return `AttributeValue`.
If it receives undefined and `removeUndefinedValues` is not set, it should throw. That's why pre-processing was added in map/set/list.

I can change `convertToAttr` return value to `AttributeValue | undefined` and perform post-processing in map/reduce operations to remove undefined values being returned.

The latter is better, as removeUndefinedValues check would be limited to `convertToScalarAttr` function.

Comment (trivikr) [reply packages/util-dynamodb/src/convertToAttr.ts:26]: Update: Returning `AttributeValue | undefined` from `convertToAttr` makes the implementation more complex as `convertToAttr` is exported.

Experimental [diff.patch.txt](https://github.com/aws/aws-sdk-js-v3/files/5737466/diff.patch.txt) attached.

Comment (AllanZhengYP) [reply packages/util-dynamodb/src/convertToAttr.ts:26]: I see. Thank you a lot for putting this up!

**Review by AllanZhengYP — APPROVED** [review]

**Pull request merged**